addons/l10n_pt_pos/models/pos_order.py

```diff
@@ -0,0 +1,220 @@
+            and o.l10n_pt_pos_inalterable_hash
+            and not o.l10n_pt_pos_qr_code_str  # Skip if already computed
+        )):
+            L10nPtHashingUtils._l10n_pt_verify_qr_code_prerequisites(order.company_id, order.l10n_pt_pos_atcud)
+
+            company_vat = re.sub(r'\D', '', order.company_id.vat)
+            partner_vat = re.sub(r'\D', '', order.partner_id.vat or '999999990')
+            details_by_tax_group = get_details_by_tax_category(order)
+            tax_letter = 'I'
+            if order.company_id.l10n_pt_account_region_code == 'PT-AC':
+                tax_letter = 'J'
+            elif order.company_id.l10n_pt_account_region_code == 'PT-MA':
+                tax_letter = 'K'
+
+            qr_code_str = ""
+            qr_code_str += f"A:{company_vat}*"
+            qr_code_str += f"B:{partner_vat}*"
+            qr_code_str += f"C:{order.partner_id.country_id.code if order.partner_id and order.partner_id.country_id else 'Desconhecido'}*"
+            qr_code_str += "D:FR*"
+            qr_code_str += "E:N*"
+            qr_code_str += f"F:{format_date(self.env, order.date_order, date_format='yyyyMMdd')}*"
+            qr_code_str += f"G:{order.l10n_pt_pos_document_number}*"
+            qr_code_str += f"H:{order.l10n_pt_pos_atcud}*"
+            qr_code_str += f"{tax_letter}1:{order.company_id.l10n_pt_account_region_code}*"
+            if details_by_tax_group.get('E'):
+                qr_code_str += f"{tax_letter}2:{details_by_tax_group.get('E')['base']}*"
+            for i, tax_category in enumerate(('R', 'I', 'N')):
+                if details_by_tax_group.get(tax_category):
+                    qr_code_str += f"{tax_letter}{i * 2 + 3}:{details_by_tax_group.get(tax_category)['base']}*"
+                    qr_code_str += f"{tax_letter}{i * 2 + 4}:{details_by_tax_group.get(tax_category)['vat']}*"
+            qr_code_str += f"N:{format_amount(order, order.amount_tax)}*"
+            qr_code_str += f"O:{format_amount(order, order.amount_total)}*"
+            qr_code_str += f"Q:{order.l10n_pt_pos_inalterable_hash_short}*"
+            qr_code_str += "R:0000"  # TODO: Fill with Certificate number provided by the Tax Authority
+            order.l10n_pt_pos_qr_code_str = qr_code_str
+
+    def _l10n_pt_pos_get_integrity_hash_fields(self):
+        if self.company_id.account_fiscal_country_id.code != 'PT':
+            return []
+        return ['date_order', 'create_date', 'amount_total', 'name']
+
+    def _l10n_pt_pos_get_sequence_info(self):
+        self.ensure_one()
+        sequence_prefix = re.sub(r'[^A-Za-z0-9]+', '_', '_'.join(self.name.split('/')[:-1])).rstrip('_')
+        sequence_number = self.name.split('/')[-1]
+        return sequence_prefix, sequence_number
+
+    def _l10n_pt_pos_hash_compute(self, previous_hash=None):
+        if self.company_id.account_fiscal_country_id.code != 'PT' or not self._context.get('l10n_pt_force_compute_signature'):
```

Comment:
I know it is used in tests, but I don't understand why it is needed and it is not possible to just set the right country?